### ISSUE TITLE
Sampler questions

### ISSUE THREAD

**Issue by cambel** [astooke/rlpyt#64]
Hi, I have been testing this repo with my own environment.
For now, I am just testing a very simple 2 degrees of freedom motion with a robotic arm.
I am using the SAC algorithm based on example_2 and it kind of works but still I don't really understand how the overall execution works.

I am using the SerialSampler, but it is required to have at least 1 evaluation environment (eval_n_envs), how does that work, why are at least 2 environments needed? One for training and one for evaluation?

I am trying to understand because I would like to save a learned model so that I can reuse it later.

**Comment by wiekern** [comment]
having this question as well. I think u could create a copy eval env totally same to the training env at this moment, at least it satisfies your needs.

**Comment by astooke** [comment]
Hi!  Yes, the evaluation envs and training envs are separate objects.  There are a few reasons for this, 1) the state of the training and eval envs are separate and not interrupted by switching to the other, 2) you might want to run the eval env with different settings, 3) you might want a different number of training envs vs eval envs, for algorithm considerations and running speed.

But you can just pass the same `eval_env_kwargs` as `env_kwargs`, and they will be built the same.  They are instantiated separately, so you don't have to worry about your env being able to be copied.

To save a learned agent for later, you can used the `snapshot_mode` argument to the `logger_context()`, which gets passed into the logger.  `snapshot_mode=last` will save the newest agent as training progresses (overwriting the previously saved one).

Hope this helps!

**Comment by cambel** [comment]
@astooke Thanks for your reply.

Actually, for now, I just wanted to train without and evaluation environment, and I found that there is a class MinibatchRl that allows me to do that. But it is great that we can have both kinds of environment separately.

Great, I didn't notice that snapshot_mode option. I had already made a different way to save the agent. If a save the snapshot, is there already an option to also load a saved snapshot?
Or should I just manually load the file (torch.load()) and then pass the agent_state_dict to the agent's argument initial_model_state_dict?
